src/bson.cc

```diff
@@ -337,11 +337,14 @@ void BSONSerializer<T>::SerializeValue(void *typeLocation,
   // Process all the values
   if (value->IsNumber()) {
     double doubleValue = NanTo<double>(value);
-    int intValue = (int)doubleValue;
-    if (intValue == doubleValue) {
+    double flooredValue = floor(doubleValue);
+    if (flooredValue == doubleValue && flooredValue <= INT32_MAX && flooredValue >= INT32_MIN) {
```

Comment:
The previous code didn't handle this either, but what's the right behavior for -0? With the JS version we get the same behavior, but I'm not sure if it's the "correct" one:

```js
> bson.deserialize(bson.serialize({ a: -0 }))
{ a: 0 }
> bson.deserialize(bson.serialize({ a: bson.Double(-0) }))
{ a: -0 }
```

Reply:
Very good point, thanks for bringing it up, I propose we align with js-bson and serialize to int32 for consistency, its least surprising I suppose. The Double class as a wrapper is still an avenue that can be used to preserve this edge case, until a future where we consider if this is the correct choice.

Reply:
To be clear, this was also what this code was doing before, but I do like that we're being explicit about it now 👍